core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java

```diff
@@ -0,0 +1,199 @@
+    return new BlockPageId(blockId, pageIndex, INVALID_BLOCK_SIZE);
+  }
+
+  /**
+   * @param blockId
+   * @param blockSize
+   * @return file ID
+   */
+  public static String fileIdOf(long blockId, long blockSize) {
+    return String.format(FILE_ID_TEMPLATE, blockId, blockSize).intern();
+  }
+
+  /**
+   * @param blockId
+   * @return file ID
+   */
+  public static String tempFileIdOf(long blockId) {
+    return fileIdOf(blockId, INVALID_BLOCK_SIZE);
+  }
+
+  /**
+   * @param fileId
+   * @return block ID
+   */
+  public static long parseBlockId(String fileId) {
+    Matcher matcher = FILE_ID_PATTERN.matcher(fileId);
+    if (matcher.matches()) {
+      try {
+        // block id can be negative, Long.parseLong will throw in this case
+        return Long.parseUnsignedLong(matcher.group(1), 16);
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException(
+            String.format("fileId %s does not contain a valid block ID", fileId), e);
+      }
+    }
+    throw new IllegalArgumentException(
+        String.format("fileId %s is not a valid paged block ID", fileId));
+  }
+
+  /**
+   * @param fileId
+   * @return block size
+   */
+  public static long parseBlockSize(String fileId) {
+    Matcher matcher = FILE_ID_PATTERN.matcher(fileId);
+    if (matcher.matches()) {
+      try {
+        String blockSizeString = matcher.group(2);
+        long size = Long.parseLong(blockSizeString, 16);
+        Preconditions.checkArgument(size >= 0, "negative block size: %s", blockSizeString);
+        return size;
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException(
+            String.format("fileId %s does not contain a valid block size", fileId), e);
+      }
+    }
+    throw new IllegalArgumentException(
+        String.format("fileId %s is not a valid paged block ID", fileId));
+  }
+
+  /**
+   * @return the block ID
+   */
+  public long getBlockId() {
+    return mBlockId;
+  }
+
+  /**
+   * @return the block size
+   */
+  public long getBlockSize() {
+    return mBlockSize;
+  }
+
+  /**
+   * Try to downcast a {@link PageId} to a {@link BlockPageId}. If the object is already a block
+   * page ID, then it is immediately returned. Otherwise, attempt to parse the {@code fileId} of
+   * the page if the encoded file name matches a block and contains all necessary metadata.
+   *
+   * @param pageId the page ID to downcast
+   * @return the downcast block page ID
+   * @throws IllegalArgumentException if the page ID cannot be cast to a block page ID
+   */
+  public static BlockPageId downcast(PageId pageId) {
+    if (pageId instanceof BlockPageId) {
+      return (BlockPageId) pageId;
+    }
+    String fileId = pageId.getFileId();
+    try {
+      long blockId = parseBlockId(fileId);
+      long blockSize = parseBlockSize(fileId);
+      return new BlockPageId(blockId, pageId.getPageIndex(), blockSize);
+    } catch (IllegalArgumentException e) {
```

Comment:
It's a bit weird that we catch and throw the same exception here... better we just bypass the IllegalArgumentException from parseBlockId/parseBlockSize